flaml/automl/task/task.py

```diff
@@ -0,0 +1,197 @@
+from abc import ABC, abstractmethod
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+
+try:
+    import ray
+except ImportError:
+    ray = None
+
+# TODO: if your task is not specified in here, define your task as an all-capitalized word
+SEQCLASSIFICATION = "seq-classification"
+MULTICHOICECLASSIFICATION = "multichoice-classification"
+TOKENCLASSIFICATION = "token-classification"
+
+SEQREGRESSION = "seq-regression"
+
+TS_FORECASTREGRESSION = (
+    "forecast",
+    "ts_forecast",
+    "ts_forecast_regression",
+)
+REGRESSION = ("regression", SEQREGRESSION, *TS_FORECASTREGRESSION)
+TS_FORECASTCLASSIFICATION = "ts_forecast_classification"
+TS_FORECASTPANEL = "ts_forecast_panel"
+TS_FORECAST = (
+    *TS_FORECASTREGRESSION,
+    TS_FORECASTCLASSIFICATION,
+    TS_FORECASTPANEL,
+)
+CLASSIFICATION = (
+    "binary",
+    "multiclass",
+    "classification",
+    SEQCLASSIFICATION,
+    MULTICHOICECLASSIFICATION,
+    TOKENCLASSIFICATION,
+    TS_FORECASTCLASSIFICATION,
+)
+RANK = ("rank",)
+SUMMARIZATION = "summarization"
+NLG_TASKS = (SUMMARIZATION,)
+NLU_TASKS = (
+    SEQREGRESSION,
+    SEQCLASSIFICATION,
+    MULTICHOICECLASSIFICATION,
+    TOKENCLASSIFICATION,
+)
+NLP_TASKS = (*NLG_TASKS, *NLU_TASKS)
+
+
+def get_classification_objective(num_labels: int) -> str:
+    if num_labels == 2:
+        objective_name = "binary"
+    else:
+        objective_name = "multiclass"
+    return objective_name
+
+
+class Task(ABC):
+    estimators = {}
```

Comment:
I see this abstract class is interesting and will allow more powerful time-series tasks.
Maybe spending some words on how that's supposed to work might be useful? 

Here's my understanding:

> Abstract base class for a machine learning task. 
Class definitions should implement abstract methods and provide a non-empty dictionary of estimator classes. 
A Task can be suitable to be used for multiple machine-learning tasks (e.g. classification or regression) or be implemented specifically for a single one depending on the generality of data validation and model evaluation methods implemented. The instantiation of a Task object stores the training data and labels.
FLAML evaluates at runtime how to behave exactly, relying on `task_name`, which describes the task family.

Reply:
Thanks! It's a good point and very helpful of you to include a docstring for me :D. I'll make that change today and add docstrings throughout the new ABC

Reply:
Great! I am happy I got the main intent and thank you for your response.

Still, I had not a full understanding of the following points when I wrote that:
1. why X,y are necessary at the initialization,
2. the fact that the `Task` class is a different concept wrt the "machine learning term to designate a family of problems" (full awareness came later).

might be the opportunity to improve my docstring draft a bit :)

Reply:
Thanks for the feedback. I've used your docstring as a start and modified to hopefully answer the additional questions listed